### PR TITLE
feature: waypoint status CLI

### PR DESCRIPTION
**Pull request by briancain**
This pull request introduces a new command `waypoint status`.

- [ ] Future PR: `waypoint status check`

Fixes https://github.com/hashicorp/waypoint/issues/1501

### CODE REVIEW

File: internal/cli/status.go
Comment (denysvitali) [inline internal/cli/status.go:159]: I would suggesting using `nil`. Althogh rare, I can see that somebody could eventually use `???` as a workspace and break this part 😅

Comment (briancain) [reply internal/cli/status.go:159]: Definitely! Good call.

Comment (krantzinator) [reply internal/cli/status.go:159]: Hm, this seems like odd behavior, in that after an `init` we should know the workspace? And in fact we know it if we specify the app.

```
$ waypoint status
Current status for project "example-go" in server context "localhost:9701".

APP       	WORKSPACE	LATEST STATUS	LAST CHECK 
example-go	default  	N/A          	          	

...
$ waypoint status -app=example-go
Current status for application "example-go" in project "example-go" in server context "localhost:9701".

Application Summary
APP       	WORKSPACE	LATEST STATUS	LAST CHECK 
example-go	???      	N/A          	          	
```

File: internal/cli/status.go
Comment (krantzinator) [inline internal/cli/status.go:128]: UX nit -- seeing both of these marked as an error makes me as a user think two errors happened, so I was confused for half a second. Can we combine them like `c.ui.Output("CLI failed to format app status: "+err.Error(), terminal.WithErrorStyle())`

File: internal/cli/status.go
Comment (krantzinator) [inline internal/cli/status.go:76]: We do allow projects with `/` in the name. I would support removing this ability and requiring the user to specify the app via the `-app` flag, rather than having two ways of doing that.

File: internal/cli/status.go
Comment (krantzinator) [inline internal/cli/status.go:116]: Another combo place!

File: internal/cli/status.go
Comment (krantzinator) [inline internal/cli/status.go:456]: I think this flow needs a bit more work. The workspace we specify via the global `workspace` flag is not reflected.
```
$ waypoint status -workspace=two
Current status for project "example-python" in server context "localhost:9701".

APP           	WORKSPACE	LATEST STATUS	LAST CHECK     
example-python	default  	✔ READY      	25 minutes ago	
```

File: internal/cli/filter_status.go
Comment (mitchellh) [inline internal/cli/filter_status.go:130]: should this be `opts ==` too? Or did we mean for this to actually be a bitmask?

If its a bitmask we need to change our constant `iota` above to `1 << iota` I think.

File: internal/cli/status.go
Comment (briancain) [reply internal/cli/status.go:159]: Good catch @krantzinator ! That behavior was from me trying to be helpful if worksace was unset and just "try" default. But I think we can rely on the API here to do the right thing if we pass it as unset so I'll remove it.

File: internal/cli/filter_status.go
Comment (briancain) [reply internal/cli/filter_status.go:130]: Oops, this needs to be the same change I made with the other if-statements and I had just missed it! I think the original intention was to use a bitmask but checking against `opts` like the change I made above works as intended too.

File: internal/cli/status.go
Comment (denysvitali) [inline internal/cli/status.go:442]: This part is repeated a couple of times. I guess it's a good candidate for a function

File: internal/cli/status.go
Comment (izaaklauer) [inline internal/cli/status.go:262]: Same problem here - I think you need more than just the most recent deploy or release status report to determine overall app status.

One piece of WP-075 is building APIs that get you a deployment/release with it's corresponding latest status report, which I think would be useful here. I've been holding off building that until WP-080 stabilizes, but I think it's just what you need here.

**Review by izaaklauer — CHANGES_REQUESTED** [review]
Nice! I'm really looking forward to this feature 🤩. 

It might be worth adding screenshots of the CLI output to the description.

File: internal/cli/status.go
Comment (briancain) [reply internal/cli/status.go:169]: That's expected in this case I think. The column we're displaying for this app is the _latest status_ check recorded rather than an aggregate of its status history.

File: internal/cli/status.go
Comment (briancain) [reply internal/cli/status.go:345]: Yep good call, I removed these ids, I don't think they were very useful! I'm just showing sequence id in the way `deployment list` shows it now (i.e. `v<sequence_id>`)

File: internal/cli/status.go
Comment (briancain) [reply internal/cli/status.go:325]: I might as well update it to be lifecycle! Probably matches closer to our docs on what an operation is too here.

File: internal/cli/status.go
Comment (briancain) [reply internal/cli/status.go:331]: I felt like for certain headers passing in the string slice directly ran off the screen and wasn't too readable so I kept them separate for now.

File: internal/cli/status.go
Comment (briancain) [reply internal/cli/status.go:261]: Since this is the most detailed view, I could basically grab the status report associated with the latest deployment and or release (if a release happened) and show them both here in separate columns.

File: internal/cli/status.go
Comment (izaaklauer) [reply internal/cli/status.go:168]: Hmm, this deviates from the logic in the RFC:

> For the top level status command, we want to let users know the health of a project given all of its applications health. When we determine the value of this column, we’ll need to look through each application's overall health status and show the total statuses for each app. We will omit any status state that isn’t covered. For example, if we have 3 apps, 2 of which are READY and 1 is error, we should show “2 READY 1 ERROR” for the overall aggregate app status.

I think I much prefer the RFC logic. As an app developer, I don't think this field is useful to me if it's flapping between new and old deployments and releases. I would expect this to be some kind of view into the health of my application.

With the bundle APIs introduced in https://github.com/hashicorp/waypoint/pull/1994, it should be pretty easy to get the latest status report for the latest deploy, the latest status report for the latest release, and then either display them both, or one combined status that can be `PARTIAL`.

File: internal/cli/status.go
Comment (izaaklauer) [inline internal/cli/status.go:384]: You're putting `Name` in the `Type` column here.

**Review by izaaklauer — CHANGES_REQUESTED** [review]

File: internal/cli/status.go
Comment (briancain) [reply internal/cli/status.go:322]: Fixed that, I think platform might still be useful and I don't mind duplicating the info too much. Especially in the json view.

File: internal/cli/status.go
Comment (briancain) [reply internal/cli/status.go:394]: Fixed!

File: internal/cli/status.go
Comment (briancain) [reply internal/cli/status.go:299]: Agreed. I will resolve these for now, because they should all be fixed once the UI methods is merged.

File: internal/cli/status.go
Comment (briancain) [reply internal/cli/status.go:865]: This whole method will be deleted once the UI methods are available so I'll go ahead and resolve this.

File: internal/cli/status.go
Comment (briancain) [reply internal/cli/status.go:168]: We can just display them both. I'll wait to update it once the UI bundle methods are available.

File: internal/cli/status.go
Comment (briancain) [reply internal/cli/status.go:622]: I think I'd rather just show both deployment and release statuses as two separate columns in the project view. Then we don't have to guess and people can interpret them separately rather than trying to combine deployment and releases statuses as 1 health report. This will also match the rest of the views too.

**Review by krantzinator — APPROVED** [review]
Looks great! Minor nit on formatting of the `waypoint status -app=APP-NAME` result: the `VERSION` and `WORKSPACE` words a bit close together and look like they are the same column. But that might be the responsibility of our cli formatter?
image

**Comment by briancain** [comment]
> But that might be the responsibility of our cli formatter?

Correct...we use [tablewriter](https://github.com/olekukonko/tablewriter) to format this, which is the same package that's used for `waypoint deployment list`. Agreed that the spacing it sometimes chooses is odd 😞

**Review by izaaklauer — APPROVED** [review]

**Pull request merged**
